zebra-network/src/meta_addr.rs

```diff
@@ -103,51 +410,407 @@ impl PartialOrd for PeerAddrState {
+                Failed {
+                    // When the attempt message arrives, it will replace this default.
+                    // (But it's a reasonable default anyway.)
+                    last_attempt: old_last_attempt.unwrap_or_else(Utc::now),
+                    last_success: old_last_success,
+                    last_failed: Utc::now(),
+                    untrusted_last_seen: old_untrusted_last_seen,
+                    // replace old services with new services if present
+                    untrusted_services: services.or(old_untrusted_services),
+                },
+            )),
+
+            // Shutdown:
+            UpdateShutdown {
+                addr,
+                services: new_services,
+            } => {
+                match old_entry.map(|old| old.last_connection_state) {
+                    // Responded:
+                    // Keep as responded, update services if present
+                    Some(Responded {
+                        last_attempt,
+                        last_success,
+                        last_failed,
+                        untrusted_last_seen,
+                        services: old_services,
+                    }) => Some(MetaAddr::new(
+                        addr,
+                        Responded {
+                            last_attempt,
+                            last_success,
+                            last_failed,
+                            untrusted_last_seen,
+                            // this could be redundant, but it handles multiple
+                            // connections and update reordering better
+                            services: new_services.unwrap_or(old_services),
+                        },
+                    )),
+
+                    // Attempted or Failed:
+                    // Change to Failed
+                    // Update last_failed and services if present
+                    Some(AttemptPending { .. })
+                    | Some(Failed { .. })
+                    // Unexpected states: change to Failed anyway
+                    | Some(NeverAttemptedSeed)
+                    | Some(NeverAttemptedGossiped { .. })
+                    | Some(NeverAttemptedAlternate { .. })
+                    | None => Some(MetaAddr::new(
+                        addr,
+                        Failed {
+                            // When the attempt message arrives, it will replace this default.
+                            // (But it's a reasonable default anyway.)
+                            last_attempt: old_last_attempt.unwrap_or_else(Utc::now),
+                            last_success: old_last_success,
+                            last_failed: Utc::now(),
+                            untrusted_last_seen: old_untrusted_last_seen,
+                            untrusted_services: new_services.or(old_untrusted_services),
+                        },
+                    )),
+                }
+            }
+        }
+    }
```

Comment:
In general, changes update the relevant field to `now`, but keep other fields the same.

For example:
* An attempted change sets the attempted field to `now`
* A success change sets the success field to `now`, and also sets the attempted field if it was `None`

Reply:
If there is a bug in the updates, that could be causing issues with the peer set.
But that seems unlikely, because of the initial seed handshakes.

Maybe we're filling up a channel somewhere?